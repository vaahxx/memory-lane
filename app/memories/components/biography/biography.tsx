"use client";

import { PencilSquareIcon } from "@heroicons/react/24/outline";
import { useState } from "react";
import { Button } from "../../../components/button/button";
import { useParams } from "next/navigation";
import { BiographyForm } from "@/app/components/forms/biography-form";

type BiographyProps = {
  text: string;
};

export function Biography({ text }: BiographyProps) {
  const [isEditing, setIsEditing] = useState<boolean>(false);
  const { id } = useParams();

  return (
    <div className='flex bg-white shadow p-6 rounded-lg gap-8 w-full'>
      <div className='flex items-center justify-between w-full'>
        {!isEditing ? (
          <>
            <p className='text-lg whitespace-pre-line'>{text}</p>
            <Button onClick={() => setIsEditing(true)} variant='transparent'>
              <PencilSquareIcon className='h-6 w-7 flex hover:stroke-green-400 transition-all ease-in duration-75' />
            </Button>
          </>
        ) : (
          <BiographyForm
            biography={text}
            userId={Number(id)}
            onSubmit={() => setIsEditing(false)}
            onCancel={() => setIsEditing(false)}
          />
        )}
      </div>
    </div>
  );
}
