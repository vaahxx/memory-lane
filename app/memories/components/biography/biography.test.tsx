import { fireEvent, render, screen } from "@testing-library/react";
import { Biography } from "./biography";
import { useFormStatus } from "react-dom";

jest.mock("react-dom", () => ({
  ...jest.requireActual("react-dom"),
  useFormState: jest.fn().mockReturnValue([{}, jest.fn()]),
  useFormStatus: jest.fn(),
}));

jest.mock("next/navigation", () => ({
  useParams: jest.fn().mockReturnValue({
    id: "123",
  }),
}));

describe("Biography", () => {
  const text = "Hello, world!";
  it("renders the text", () => {
    const { container } = render(<Biography text={text} />);

    expect(container).toMatchSnapshot();
    expect(screen.getByText(text)).toBeInTheDocument();
  });

  it("renders the button", () => {
    const { container } = render(<Biography text={text} />);

    expect(container).toMatchSnapshot();
    expect(screen.getByRole("button")).toMatchSnapshot();
  });

  it("if is editing, then biography form is rendered", () => {
    render(<Biography text={text} />);

    // Mock the return value of the useFormStatus hook
    (useFormStatus as jest.Mock).mockReturnValue({
      pending: false,
    });

    fireEvent.click(screen.getByRole("button"));

    expect(screen.getByTestId("biography-form")).toMatchSnapshot();
  });
});
