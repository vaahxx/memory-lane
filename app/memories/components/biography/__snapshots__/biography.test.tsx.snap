// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Biography if is editing, then biography form is rendered 1`] = `
<form
  class="flex flex-col gap-3 w-full"
  data-testid="biography-form"
>
  <div
    class="mb-6 flex flex-col justify-items-start"
  >
    <input
      name="user_id"
      type="hidden"
      value="123"
    />
    <label
      class="block mb-2 text-sm font-medium text-gray-900"
      for="biography"
    >
      Biography
    </label>
    <textarea
      class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-green-300 focus:border-green-300 block w-full p-2.5 h-36"
      id="biography"
      name="biography"
      placeholder="Tell us about yourself"
      required=""
    >
      Hello, world!
    </textarea>
  </div>
  <div
    class="flex justify-end gap-5"
  >
    <button
      class="flex gap-2 items-center group transition-all ease-in duration-75"
      type="button"
    >
      <span
        class="flex items-center gap-2 text-base bg-gradient-to-br group-hover:from-green-400 group-hover:to-blue-600 group-hover:bg-clip-text group-hover:text-transparent"
      >
        <span>
          Cancel
        </span>
      </span>
    </button>
    <button
      class="flex items-center justify-center py-5 px-[0.15rem] overflow-hidden text-gray-900 rounded-lg group bg-gradient-to-br from-green-400 to-blue-600 group-hover:from-green-400 group-hover:to-blue-600 hover:text-white max-h-8"
      type="submit"
    >
      <span
        class="flex items-center gap-2 text-base px-1.5 py-1.5 transition-all ease-in duration-75 bg-white rounded-md group-hover:bg-opacity-0"
      >
        <span>
          Update
        </span>
      </span>
    </button>
  </div>
</form>
`;

exports[`Biography renders the button 1`] = `
<div>
  <div
    class="flex bg-white shadow p-6 rounded-lg gap-8 w-full"
  >
    <div
      class="flex items-center justify-between w-full"
    >
      <p
        class="text-lg whitespace-pre-line"
      >
        Hello, world!
      </p>
      <button
        class="flex gap-2 items-center group transition-all ease-in duration-75"
      >
        <span
          class="flex items-center gap-2 text-base bg-gradient-to-br group-hover:from-green-400 group-hover:to-blue-600 group-hover:bg-clip-text group-hover:text-transparent"
        >
          <span>
            <svg
              aria-hidden="true"
              class="h-6 w-7 flex hover:stroke-green-400 transition-all ease-in duration-75"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16.862 4.487l1.687-1.688a1.875 1.875 0 112.652 2.652L10.582 16.07a4.5 4.5 0 01-1.897 1.13L6 18l.8-2.685a4.5 4.5 0 011.13-1.897l8.932-8.931zm0 0L19.5 7.125M18 14v4.75A2.25 2.25 0 0115.75 21H5.25A2.25 2.25 0 013 18.75V8.25A2.25 2.25 0 015.25 6H10"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`Biography renders the button 2`] = `
<button
  class="flex gap-2 items-center group transition-all ease-in duration-75"
>
  <span
    class="flex items-center gap-2 text-base bg-gradient-to-br group-hover:from-green-400 group-hover:to-blue-600 group-hover:bg-clip-text group-hover:text-transparent"
  >
    <span>
      <svg
        aria-hidden="true"
        class="h-6 w-7 flex hover:stroke-green-400 transition-all ease-in duration-75"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16.862 4.487l1.687-1.688a1.875 1.875 0 112.652 2.652L10.582 16.07a4.5 4.5 0 01-1.897 1.13L6 18l.8-2.685a4.5 4.5 0 011.13-1.897l8.932-8.931zm0 0L19.5 7.125M18 14v4.75A2.25 2.25 0 0115.75 21H5.25A2.25 2.25 0 013 18.75V8.25A2.25 2.25 0 015.25 6H10"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </span>
  </span>
</button>
`;

exports[`Biography renders the text 1`] = `
<div>
  <div
    class="flex bg-white shadow p-6 rounded-lg gap-8 w-full"
  >
    <div
      class="flex items-center justify-between w-full"
    >
      <p
        class="text-lg whitespace-pre-line"
      >
        Hello, world!
      </p>
      <button
        class="flex gap-2 items-center group transition-all ease-in duration-75"
      >
        <span
          class="flex items-center gap-2 text-base bg-gradient-to-br group-hover:from-green-400 group-hover:to-blue-600 group-hover:bg-clip-text group-hover:text-transparent"
        >
          <span>
            <svg
              aria-hidden="true"
              class="h-6 w-7 flex hover:stroke-green-400 transition-all ease-in duration-75"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16.862 4.487l1.687-1.688a1.875 1.875 0 112.652 2.652L10.582 16.07a4.5 4.5 0 01-1.897 1.13L6 18l.8-2.685a4.5 4.5 0 011.13-1.897l8.932-8.931zm0 0L19.5 7.125M18 14v4.75A2.25 2.25 0 0115.75 21H5.25A2.25 2.25 0 013 18.75V8.25A2.25 2.25 0 015.25 6H10"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;
